test/model_tests.cpp

```diff
@@ -249,6 +249,180 @@ TEST(ModelTests, BeamSearchGptCuda) {
+
+  // Complete expected sequence (input + generated) from model_qa.cpp using the working phi3-fp16-nvtrt model
+  std::vector<int32_t> expected_output{
+    32006, 887, 526, 263, 8444, 29871, 23869, 20255, 29889, 32007, 32010, 6324, 29892, 1128, 526, 366, 29973, 32007, 32001,  // Input tokens (19)
+    15043, 29991, 306, 29915, 29885, 2599 
+  };
+  auto config = OgaConfig::Create(model_path);
+  config->ClearProviders();
+  config->AppendProvider("NvTensorRtRtx");
+  auto model = OgaModel::Create(*config);
+
+  int max_length = 25;
+  int batch_size = static_cast<int>(input_ids_shape[0]);
+  auto params = OgaGeneratorParams::Create(*model);
+  params->SetSearchOption("max_length", max_length);
+  params->SetSearchOption("batch_size", batch_size);
+
+  auto generator = OgaGenerator::Create(*model, *params);
+  generator->AppendTokens(input_ids);
+
+  while (!generator->IsDone()) {
+    generator->GenerateNextToken();
+  }
+
+  // Verify outputs match expected outputs
+  for (int i = 0; i < batch_size; i++) {
+    auto sequence = generator->GetSequence(i);
+    auto* expected_output_start = &expected_output[i * max_length];
+    
+    EXPECT_TRUE(0 == std::memcmp(expected_output_start, sequence.data(), max_length * sizeof(int32_t)));
+  }
+}
+
+TEST(ModelTests, GreedySearchPhi3NvTensorRtRtx) {
+  for (auto model_path : c_phi3_nvtrt_model_paths)
+    Test_GreedySearch_Phi3_NvTensorRtRtx(model_path.first, model_path.second);
+}
+
+void Test_OutOfPlaceKvCache_Phi3_NvTensorRtRtx(const char* model_path, const char* model_label) {
+  std::vector<int64_t> input_ids_shape{1, 19};
+  std::vector<int32_t> input_ids{
+      32006, 887, 526, 263, 8444, 29871, 23869, 20255, 29889,
+      32007, 32010, 6324, 29892, 1128, 526, 366, 29973, 32007, 32001};
+
+  auto config = OgaConfig::Create(model_path);
+  config->ClearProviders();
+  config->AppendProvider("NvTensorRtRtx");
+  auto model = OgaModel::Create(*config);
+
+  int max_length = 25;
+  int batch_size = static_cast<int>(input_ids_shape[0]);
+  auto params = OgaGeneratorParams::Create(*model);
+  params->SetSearchOption("max_length", max_length);
+  params->SetSearchOption("batch_size", batch_size);
+
+  // Test 1: Basic rewind functionality
+  auto generator1 = OgaGenerator::Create(*model, *params);
+  generator1->AppendTokens(input_ids);
+
+  // Generate some tokens
+  for (int i = 0; i < 8; i++) {
+    generator1->GenerateNextToken();
+  }
+
+  auto sequence_before_rewind = generator1->GetSequence(0);
+  std::cout << "Sequence before rewind (length " << sequence_before_rewind.size() << "): ";
+  for (size_t j = 0; j < sequence_before_rewind.size(); j++) {
+    std::cout << sequence_before_rewind[j];
+    if (j < sequence_before_rewind.size() - 1) std::cout << ", ";
+  }
+  std::cout << std::endl;
+
+  // Test out-of-place KV cache functionality by rewinding to a previous position
+  // This should create new tensor allocations (out-of-place) rather than modifying existing ones
```

Comment:
Could you please explain why rewinding tests out-of-place kv cache? what does out-of-place kv cache mean?